clickhouse/datadog_checks/clickhouse/clickhouse.py

```diff
@@ -0,0 +1,292 @@
+                            metric_info.append((name, float(value), column_type))
+                        except (ValueError, TypeError):  # no cov
+                            self.log.error('Non-numeric value `%s` for metric column `%s`', value, name)
+                            break
+
+                # Only submit metrics if there were absolutely no errors - all or nothing.
+                else:
+                    for info in metric_info:
+                        metric, value, method = info
+                        getattr(self, method)(metric, value, tags=query_tags)
+
+    def collect_version(self, extra_version_parts=None):
+        version = self.execute_query_raw('SELECT version()', lambda: 'version()')[0][0]
+
+        # The version comes in like `19.15.2.2` though sometimes there is no patch part
+        version_parts = {name: part for name, part in zip(('year', 'major', 'minor', 'patch'), version.split('.'))}
+        if extra_version_parts:
+            version_parts.update(extra_version_parts)
+
+        self.set_metadata('version', version, scheme='parts', final_scheme='calver', part_map=version_parts)
+
+    def execute_query_schema_search(self, query):
+        rows = self.execute_query_raw(query.query, lambda: ', '.join(sorted(query.views)))
+
+        # Avoid repeated lookups
+        columns = query.columns
+        get_method = getattr
+
+        result = {}
+        for column, value in rows:
+            metric_data = columns.get(column)
+            if metric_data is None:
+                continue
+
+            for metric, method in metric_data:
+                if callable(method):
+                    method, computed_value = method(value)
+                    get_method(self, method)(metric, computed_value, tags=self._tags)
+                else:
+                    get_method(self, method)(metric, value, tags=self._tags)
+
+            result[column] = value
+
+        # Return all values for possible post-processing
+        return result
+
+    def execute_query_defined_schema(self, query):
+        rows = self.execute_query_raw(query.query, lambda: ', '.join(sorted(query.views)))
+
+        # Re-use column access map for efficiency
+        result = {}
+
+        for row in rows:
+            for column, value in zip(query.columns, row):
+                result[column] = value
+
+            yield result
+
+    def execute_query_raw(self, query, source, stream=False):
+        try:
+            if stream:
+                rows = self._client.execute_iter(query)
+
+                # Trigger query execution
+                try:
+                    first_row = next(rows)
+                except StopIteration:  # no cov
+                    return iter([])
+
+                return chain((first_row,), rows)
+            else:
+                return self._client.execute(query)
+        except Exception as e:
+            raise QueryExecutionError(self._error_sanitizer.clean(str(e)), source())
```

Comment:
Having `source` callable is uselessly complicated. If you want to postpone the string building, pass a list normally to the exception, and build the string there.